port/unix/omrsysinfo.c

```diff
@@ -3603,6 +3650,56 @@ readCgroupSubsystemFile(struct OMRPortLibrary *portLibrary, uint64_t subsystemFl
 	return rc;
 }
 
+/** 
+ * Return container type
+ *
+ * @param[in] portLibrary pointer to OMRPortLibrary
+ * @param[out] containerType pointer to OMRContainerType which on successful return contains the
+ *                           container type in which the process is running.
+ *
+ * @return 0 on success, otherwise negative error code
+ */
+static int32_t
+getContainerType(struct OMRPortLibrary *portLibrary, OMRContainerType *containerType)
+{
+	int32_t rc = 0;
+	OMRCgroupEntry *cgEntryList = NULL;
+	OMRCgroupEntry *temp = NULL;
+	const char *cgroupName = NULL;
+
+	Assert_PRT_true(NULL != containerType);
+
+	/* Assume we are not in container */
+	*containerType = OMR_CONTAINER_NONE;
+
+	if (isCgroupV1Available(portLibrary)) {
+		/* Read PID 1's cgroup file /proc/1/cgroup and check cgroup name for each subsystem.
+		 * If cgroup name for each subsystem points to the root cgroup "/",
+		 * then the process is not running in a container.
+		 * For any other cgroup name, assume we are in a 'docker' container.
+		 */
+		rc = readCgroupFile(portLibrary, 1, FALSE, &cgEntryList, NULL);
+		if (0 != rc) {
+			goto _end;
+		}
+
+		temp = cgEntryList;
+		do {
+			cgroupName = cgEntryList->cgroup;
+			if (strcmp(ROOT_CGROUP, cgroupName)) {
+				*containerType = OMR_CONTAINER_DOCKER;
+				break;
+			}
+			temp = temp->next;
+		} while (temp != cgEntryList);
+	}
+_end:
+	if (NULL != cgEntryList) {
+		freeCgroupEntries(portLibrary, cgEntryList);
```

Comment:
This seems allocation heavy for the information being gathered.  Can this be done without allocating and freeing the cgroup list?

Reply:
It looks possible. Let me see if I can refactor it to make it light weight.